text/3440-virt-self.md

```diff
@@ -0,0 +1,189 @@
+- Feature Name: `virt_self`
+- Start Date: `2023-05-30`
+- RFC PR: [rust-lang/rfcs#3440](https://github.com/rust-lang/rfcs/pull/3440)
+- Rust Issue: [rust-lang/rust#3440](https://github.com/rust-lang/rust/issues/3440)
+
+# Summary
+[summary]: #summary
+
+Enable virtual dispatch of trait methods.
+
+# Motivation
+[motivation]: #motivation
+
+Coming to Rust from an OOP language such as C++ one is told to favor composition over inheritance. In general, I think that's a great thing. However, lack of language features such as delegation and what this RFC proposes can make certain patterns more difficult to express than they should be.
+
+Consider the following situation. Say we have a trait `System` that models a given system.
+
+```rust
+pub trait System {
+    fn op1(&self) -> SystemOp1;
+    fn op2(&self) -> SystemOp2;
+    // etc.
+}
+```
+
+Assume that we now have a particular impl. `SystemA` which implements `op1` through a reference to `op2`. For instance,
+
+```rust
+// asume the following
+type SystemOp1 = i64;
+type SystemOp2 = i64;
+
+pub struct SystemA {
+
+}
+
+impl System for SystemA {
+    fn op1(&self) -> SystemOp1 {
+        self.op2() * 5
+    }
+    fn op2(&self) -> SystemOp2 {
+        3
+    }
+}
+```
+
+Assume we now want to have a general purpose wrapper that allows us to somehow map the result of `op2`, say. For instance,
+
+```rust
+pub struct DoubleOp2<S: System> {
+    sys: S,
+}
+
+impl<S: System> System for DoubleOp2<S> {
+    fn op1(&self) -> SystemOp1 {
+        self.sys.op1()
+    }
+    fn op2(&self) -> SystemOp2 {
+        self.sys.op2() * 2
+    }
+}
+```
+
+Clearly, this has the intended effect of changing `op2`. However, it also has the unintended effect of keeping `DoubleOp2<SystemA>::op1()` out of sync with `DoubleOp2<SystemA>::op2`. We got static dispatch when in this context virtual dispatch made more sense.
+
+Of course, in Rust, we usually associate dynamic dispatch with the `dyn` keyword. However, `dyn System` only gives a vtable for a particular impl without virtualizing any subsequent calls. In other words, calling `op1` through `DoubleOp2<SystemA> as dyn System` will still call `SystemA::op2`.
+
+Of course, this behaviour may be what is desired. But sometimes, virtualizing at depth is more appropriate. Certainly when adopting OOP patterns.
+
+When I first stumbled upon this, I wondered if I could simply take `self` as `&dyn System` (or `Arc` or `Box` equivalent, etc.). Nope.
```

Comment:
It would definitely help for quite a few use cases, yes!

Reply:
Hm. To what degree does `#![feature(arbitrary_self_types)]` obsolete or outmode this RFC? If what you want (essentially, "more control over vtables") can be built on that more ergonomically, then I think it would be most appropriate to redraft this RFC with that feature in mind.